File: src/patterns/combined/notmvc/Observable.java
package patterns.combined.notmvc;

import java.util.ArrayList;
import java.util.Iterator;


/**
 * Created by Home on 11.12.2016.
 */
public class Observable implements QuackObservable {
    ArrayList observers = new ArrayList();
    QuackObservable duck;

    public Observable(QuackObservable duck) {
        this.duck = duck;
    }

    @Override
    public void registerObserver(Observer observer) {
        observers.add(observer);
    }

    @Override
    public void notifyObservers() {
        Iterator iterator = observers.iterator();
        while (iterator.hasNext()){
            Observer observer = (Observer)iterator.next();
            observer.update(duck);
        }
    }
}

File: src/patterns/together/observer/impl/Observable.kt
package patterns.together.observer.impl

import patterns.together.observer.Observer
import patterns.together.observer.QuackObservable

class Observable(val duck: QuackObservable) : QuackObservable {
    private val observers = arrayListOf<Observer>()

    override fun registerObserver(observer: Observer) {
        observers.add(observer)
    }

    override fun notifyObservers() {
        val iterator = observers.iterator()
        while (iterator.hasNext()) {
            val observer = iterator.next()
            observer.update(duck)
        }
    }
}